### PR TITLE
C Compatibility

### PR DESCRIPTION
**Pull request by pvelesko**
The following are the changes I needed to make to get libCEED to compile with CHIP-SPV. 
Most issues stem from from texture implementation.

### CODE REVIEW

File: include/hip/spirv_hip.hh [hunk 1]
```diff
@@ -146,10 +146,15 @@ extern const __device__ __attribute__((weak)) __hip_builtin_gridDim_t gridDim;
...
 
+#ifdef __cplusplus
 extern "C" __device__ int printf(const char *fmt, ...)
     __attribute__((format(printf, 1, 2)));
 extern "C" __device__ void abort();
-
+#else
+extern __device__ int printf(const char *fmt, ...)
+    __attribute__((format(printf, 1, 2)));
+extern __device__ void abort();
```
Comment (linehill) [inline include/hip/spirv_hip.hh:156]: Is it necessary to define these for C applications? At least, the `__device__ `attribute is defunct in C mode.

File: samples/ccompat/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,7 @@
+add_custom_command(OUTPUT saxpy.o   DEPENDS ${CMAKE_SOURCE_DIR}/samples/ccompat/saxpy.hip COMMAND ${CMAKE_BINARY_DIR}/bin/hipcc.bin ${CMAKE_SOURCE_DIR}/samples/ccompat/saxpy.hip -c -g)
+add_custom_command(OUTPUT ccompat.o DEPENDS ${CMAKE_SOURCE_DIR}/samples/ccompat/ccompat.c COMMAND ${CMAKE_BINARY_DIR}/bin/hipcc.bin ${CMAKE_SOURCE_DIR}/samples/ccompat/ccompat.c -c -g)
+
+add_custom_target(ccompat DEPENDS saxpy.o ccompat.o COMMAND ${CMAKE_BINARY_DIR}/bin/hipcc.bin ./ccompat.o ./saxpy.o -o ccompat -g)
+
+add_dependencies(ccompat hipcc.bin CHIP)
+add_test(NAME ccompat COMMAND ${CMAKE_BINARY_DIR}/samples/ccompat/ccompat)
```
Comment (linehill) [inline samples/ccompat/CMakeLists.txt:7]: Style nit: missing newline at the end of file.

File: samples/ccompat/ccompat.c [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+extern __global__ void saxpy();
+
+  dim3 blocks 	    = {1,1,1};
+  dim3 threads      = {1,1,1};
+
+int main(void) {
+  hipLaunchKernel((const void*)(&saxpy), blocks, threads, 0, 0, 0);
+  hipDeviceSynchronize();
+  return 0;
+}
```
Comment (linehill) [inline samples/ccompat/ccompat.c:15]: Style nit: missing newline.

File: samples/ccompat/saxpy.hip [hunk 4]
```diff
@@ -0,0 +1,5 @@
+#include <hip/hip_runtime.h>
+
+extern "C" {
+    __global__ void saxpy() { printf("PASSED\n"); }
+}
```
Comment (linehill) [inline samples/ccompat/saxpy.hip:5]: Style nit: missing newline.

File: src/CHIPBindings.cc [hunk 5]
```diff
@@ -3777,7 +3777,7 @@ hipError_t hipLaunchKernel(const void *HostFunction, dim3 GridDim,
                            hipStream_t Stream) {
   CHIP_TRY
   CHIPInitialize();
-  NULLCHECK(HostFunction, Args);
+  NULLCHECK(HostFunction);
```
Comment (linehill) [inline src/CHIPBindings.cc:3780]: Seems unrelated to the patch's scope.

File: samples/ccompat/ccompat.c [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#include <stdio.h>
+#include "hip/hip_runtime.h"
+#include "hip/hip_runtime_api.h"
+
+
+extern __global__ void saxpy();
+
+  dim3 blocks 	    = {1,1,1};
```
Comment (linehill) [inline samples/ccompat/ccompat.c:8]: Should we follow LLVM style here?

**Review by linehill — APPROVED** [review]
Some style nits. Otherwise, LGTM.

**Review by Sarbojit2019 — APPROVED** [review]
